go-controller/pkg/node/egressfirewall_dns/egressfirewall_dns.go

```diff
@@ -0,0 +1,278 @@
+	e.dnsEntries[dnsName].dnsResolves = ips
+
+	//update the nodeDNSInfo object so the master can coordinate all the nodes ip addresses
+	// if the node has not reported any DNS addresses before create the nodeDNSObject
+	nodeDNSInfo, err := e.wf.GetNodeDNSInfo(e.nodeName)
+	existed := true
+	if errors.IsNotFound(err) {
+		// the object *MOST LIKELY* has not been created yet, so start making it
+		nodeDNSInfo = &nodednsinfoapi.NodeDNSInfo{
+			ObjectMeta: metav1.ObjectMeta{Name: e.nodeName},
+			Spec:       nodednsinfoapi.NodeDNSInfoSpec{},
+		}
+		existed = false
+	} else if err != nil {
+		return err
+	}
+	var ipStrings []string
+	for _, ip := range e.dnsEntries[dnsName].dnsResolves {
+		ipStrings = append(ipStrings, ip.String())
+	}
+
+	if nodeDNSInfo.Spec.DNSEntries == nil {
+		nodeDNSInfo.Spec.DNSEntries = make(map[string]nodednsinfo.DNSEntry)
+	}
+
+	dnsEntry := nodeDNSInfo.Spec.DNSEntries[dnsName]
+	dnsEntry.IPAddresses = ipStrings
+	nodeDNSInfo.Spec.DNSEntries[dnsName] = dnsEntry
+
+	//create or update NodeDNSInfo
+	if existed {
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			return e.k.UpdateNodeDNSInfo(nodeDNSInfo)
+		})
+	} else {
+		_, err = e.k.CreateNodeDNSInfo(nodeDNSInfo)
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//Update wraps the dnsUpdate so that the functions can be mocked for testing
+// make a call to the backend DNS implementation to Update the resolved IPs
+// and and TTL for that address
+func (e *EgressDNS) Update(dns string) (bool, error) {
+	return e.dns.Update(dns)
+}
+
+// Run spawns a goroutine that handles updates to the dns entries for dnsNames used in
+// EgressFirewalls. The loop runs after receiving one of two signals
+// 1. a new dnsName has been added and a signal is sent to add the new DNS name, if an
+//    EgressFirewall uses a DNS name already added by another egressFirewall the previous
+//    entry is used
+// 2. If the defaultInterval has run (30 min) without updating the DNS server is manually queried
+func (e *EgressDNS) Run(defaultInterval time.Duration) {
+	var dnsName string
+	var ttl time.Time
+	var timeSet bool
+	// initially the next DNS Query happens at the default interval
+	durationTillNextQuery := defaultInterval
+	go func() {
+		for {
+			// Wait for the given duration or until something gets added
+			select {
+			case dnsName := <-e.added:
+				if err := e.dns.Add(dnsName); err != nil {
+					utilruntime.HandleError(err)
+				}
+				if err := e.updateEntryForName(dnsName); err != nil {
+					utilruntime.HandleError(err)
+				}
+			case <-time.After(durationTillNextQuery):
+				if len(dnsName) > 0 {
+					if _, err := e.Update(dnsName); err != nil {
+						utilruntime.HandleError(err)
+					}
+					if err := e.updateEntryForName(dnsName); err != nil {
+						utilruntime.HandleError(err)
+					}
+				}
+			case dnsName := <-e.removed:
+				// when there is a dnsName removed the query time should be updated
+				e.dns.Delete(dnsName)
```

Comment:
Can this fail? Shouldn't we remove it from the CRD?

Reply:
the delete here is just removing the dns entry from the underlying dns resolver I might be able to move where this happens.